engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodImportResolver.java

```diff
@@ -0,0 +1,139 @@
+  private final TopLevelScope topScope;
+
+  private InvokeMethodImportResolver(Module module, TopLevelScope topScope) {
+    this.module = module;
+    this.topScope = topScope;
+  }
+
+  @Override
+  protected String nameForImport(UnresolvedSymbol name) {
+    var fqn = module.getName().pathAsJava();
+    String moduleName;
+    if (fqn.size() == 2 && "Main".equals(module.getName().item())) {
+      moduleName = module.getPackage().libraryName().toString();
+    } else {
+      moduleName = module.getName().toString();
+    }
+    var subModuleName = moduleName + "." + name.getName();
+    return subModuleName;
+  }
+
+  @Override
+  protected List<String> partsForImport(UnresolvedSymbol imp) {
+    return module.getName().createChild(imp.getName()).createChild("any").pathAsJava();
+  }
+
+  @Override
+  protected String nameForExport(Object ex) {
+    throw new AssertionError("not used: " + ex);
+  }
+
+  @Override
+  protected String nameForType(Type e) {
+    return e.getName();
+  }
+
+  @Override
+  protected List<Object> exportsFor(Module module, String impName) {
+    return Collections.emptyList();
+  }
+
+  @Override
+  protected boolean isAll(Object ex) {
+    return false;
+  }
+
+  @Override
+  protected List<String> onlyNames(Object ex) {
+    return null;
+  }
+
+  @Override
+  protected List<String> hiddenNames(Object ex) {
+    return null;
+  }
+
+  @Override
+  protected List<Type> definedEntities(UnresolvedSymbol name) {
+    var associatedType = module.getScope().getType(name.getName());
+    var allRelativeTypes = module.getScope().getTypes().values();
+    return Stream.concat(associatedType.stream(), allRelativeTypes.stream()).toList();
+  }
+
+  @Override
+  protected Module loadLibraryModule(LibraryName libraryName, String moduleName)
+      throws IOException {
+    var optionModule = topScope.getModule(moduleName);
+    return optionModule.isPresent() ? optionModule.get() : null;
+  }
+
+  @Override
+  protected EnsoObject createResolvedImport(UnresolvedSymbol imp, List<Object> exp, Module m) {
+    return m.getScope().getAssociatedType();
+  }
+
+  @Override
+  protected EnsoObject createResolvedType(UnresolvedSymbol imp, List<Object> exp, Type typ) {
+    return typ;
+  }
+
+  @Override
+  protected EnsoObject createErrorPackageCoundNotBeLoaded(
+      UnresolvedSymbol imp, String impName, String loadingError) {
+    return null;
+  }
+
+  @Override
+  protected EnsoObject createErrorModuleDoesNotExist(UnresolvedSymbol imp, String impName) {
+    return null;
+  }
+
+  @CompilerDirectives.TruffleBoundary
+  private static boolean isNamedAsAssociatedType(Type t) {
+    var at = t.getDefinitionScope().getAssociatedType();
+    var byType = at == t;
+    if (byType) {
+      return true;
+    }
+    var byName = at.getName().equals(t.getName());
+    if (!byName) {
+      return false;
+    }
+    var atq = at.getQualifiedName();
+    var tq = t.getQualifiedName();
+    var tqParentOption = tq.getParent();
```

Comment:
introducing those extra vars just to name them in a cryptic way is kind of confusing and not particularly usefull

Reply:
`tqPatternOption` is referenced twice on the next line. Giving the variable a name is standard pattern to avoid boilerplate. Áh, you mean: I should use `map` and lambda function - no, thank you - I don't want to use lambda functions in these simple situations.